### PR TITLE
[Feature] Add Qwen3 MoE CUDA AFD support

### PR DESCRIPTION
**Pull request by sinpeyw**
## Purpose

Add CUDA AFD support for vLLM 0.26.0's native `Qwen3MoeForCausalLM` architecture, covering:

- `Qwen/Qwen3-30B-A3B`
- `Qwen/Qwen3-30B-A3B-Instruct-2507`
- `Qwen/Qwen3-30B-A3B-Thinking-2507`
- `Qwen/Qwen3-Coder-30B-A3B-Instruct`
- `Qwen/Qwen3-235B-A22B`

The adapter splits at the native complete-MLP boundary, keeps routing and experts together on FFN, and preserves the native Qwen forward and weight-loading lifecycle. It does not reuse the Qwen3.5/3.6 hybrid-state and router-logits transport introduced by the separate work discussed in #179 and draft PR #181.

## Issue

- Related issue: #215
- Closing keyword: none; keep the RFC open through implementation review

## Scope

- In scope:
  - CUDA, single-host vLLM V1, and synchronous `P2pNcclAFDConnector`;
  - AFD-only Qwen3 MoE registration and role-aware module construction;
  - one-pass role-filtered checkpoint loading through the native loader;
  - eager, `FULL_DECODE_ONLY` CUDA graph, and native two-ubatch DBO;
  - 1A1F, 2A2F, 2A1F, TP=1/2, and a 235B 4A4F/TP4/EP4 validation profile.
- Out of scope:
  - Qwen3.5/3.6, dense Qwen3, Qwen3-Next, and Qwen3-VL architectures;
  - NPU, ROCm, MUSA, multi-node, and asynchronous connectors;
  - Attention-side routing or a new connector/wire payload;
  - sequence-parallel MoE, EPLB, pipeline parallelism, speculative decoding, MTP/Eagle, LoRA, quantized checkpoints, long-context, or performance claims.

## Implementation Notes

- Register `AFDQwen3MoeForCausalLM` under an AFD-only alias; native `Qwen3MoeForCausalLM` registration remains unchanged.
- Inject `AFDQwen3MoeDecoderLayer` through the native `Qwen3MoeModel.decoder_layer_type` hook and inherit native decoder/model forward behavior.
- Attention constructs native Attention and norms plus the existing parameter-free `RemoteFFNProxy`; FFN constructs the complete native dense MLP or `Qwen3MoeSparseMoeBlock`.
- Filter checkpoint paths once by role, retaining native parameter names, packed mappings, TP/EP behavior, and `AutoWeightsLoader` results.
- No vLLM source file, connector contract, communication payload, FusedMoE, MoERunner, quantization path, or kernel is changed.

## Test Plan

Static and CPU contracts:

```bash
uv run ruff check .
uv run ruff format --check .
git diff --check
uv run pytest -q tests/unit
```

The GPU validation evidence below was collected before the checked-in E2E harness was split from this model-only PR. It covers a 10-case matrix with eager, 64-request/concurrency-32 pressure, CUDA graph, DBO, TP=2, 2A2F, and 2A1F. Additional controlled checks compare native TP1 with AFD 1A1F on Qwen3-30B-A3B BF16 and exercise the 235B profile as 4A4F/TP4/EP4.

## Test Result

- GPU validation source: `b5e91a8184af9e45e05d1175c3485d239a32625b`
- Rebased model-only source: `0b9d2dd856b57ca999074be7104d77686f6fd84c`
- Base: `upstream/main@18197b3b36aed7868cfac5b092545d862140b945`

Environment:

- Python 3.12.12
- PyTorch 2.11.0+cu129
- vLLM 0.26.0
- 8 x NVIDIA A800-SXM4-80GB, driver 550.54.14

| Gate | Result |
| --- | --- |
| Ruff check / format / diff check | PASS |
| Full CPU unit suite on the rebased model-only source | PASS |
| Focused registration, construction, and weight contracts | PASS, 47/47 |
| 30B GPU validation matrix | PASS, 10/10 for Base with the strengthened runtime assertions; Instruct-2507, Thinking-2507, and Coder identities passed the earlier pre-review 10-case matrix |
| Qwen3-30B-A3B BF16 oracle | PASS for English, Chinese, code, and arithmetic prompts; identical generated token IDs and top-5 membership, maximum sampled/top-5 logprob error `0.0` |
| Pressure | PASS, 64/64 requests at concurrency 32 |
| Qwen3-235B-A22B execution | PASS, 4A4F/TP4/EP4 eager, four concurrent requests |
| Cleanup | PASS, no AFD/vLLM worker, listening port, or test-owned GPU allocation remained |

This is correctness and execution evidence, not a throughput or production-readiness claim.

## Docs Impact

- Updated `README.md` model support and known limitations.
- Updated model-integration documentation.

---

<details>
<summary>Essential PR Checklist</summary>

- [x] Purpose is clear and linked to public context.
- [x] Scope is bounded.
- [x] Compatibility with vLLM v0.26.0 is considered.
- [x] No changes are made to the vLLM source checkout.
- [x] Plugin-owned classes and explicit dotted class paths are used.
- [x] Imports remain CPU-safe; CUDA execution is GPU-gated.
- [x] Rebased CPU and pre-rebase GPU validation evidence is attached.
- [x] Documentation impact is stated.

</details>

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
is there any perf comparison and profiler for compute and network in A/F

**Comment by jiangkuaixue123** [comment]
@yujuancao07 Maybe you can help review the change of e2e-test.

File: afd_plugin/v1/worker/attention_model_runner.py [hunk 1]
```diff
@@ -264,24 +274,39 @@ def _build_attention_metadata(
...
-            logits_indices,
-            use_spec_decode,
-            for_cudagraph_capture,
-            num_scheduled_tokens,
-            cascade_attn_prefix_lens,
-            slot_mappings,
-        )
+        disabled_metadata_builders = {}
+        try:
+            if ubatch_slices is not None and len(ubatch_slices) > 1:
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/attention_model_runner.py:290]: Could you clarify why this change to `AFDAttentionModelRunner` is necessary? Is disabling `supports_update_block_table` a general DBO compatibility issue introduced by upgrading to vLLM 0.26.0, or is it specific to Qwen3 MoE? If it affects AFD DBO generally, would it make sense to handle it as a separate compatibility fix rather than coupling it to the Qwen3 MoE integration?

**Review by jiangkuaixue123 — COMMENTED** [review]
Question about the `AFDAttentionModelRunner` compatibility change.

**Comment by sinpeyw** [comment]
> is there any perf comparison and profiler for compute and network in A/F

This PR currently focuses on correctness and does not make performance claims. I’ll follow up with a performance comparison and A/F compute and network profiling results.

File: afd_plugin/v1/worker/attention_model_runner.py (hunk 1 above)
Comment (sinpeyw) [reply afd_plugin/v1/worker/attention_model_runner.py:290]: You’re right. This is a general vLLM 0.26 DBO metadata-cache issue exposed by Qwen3/FlashAttention, not Qwen3 adapter logic. I’ll move the workaround and its tests into a separate compatibility PR and keep #216 focused on Qwen3 support.

Comment (sinpeyw) [reply afd_plugin/v1/worker/attention_model_runner.py:290]: I’ve moved the compatibility fix into a standalone bugfix PR, #217, with the trade-off against directly backporting vLLM #48659 documented there. I’ll remove the workaround and its generic tests/docs from #216 so this PR remains focused on Qwen3 support, then rebase it after #217 lands

Comment (sinpeyw) [reply afd_plugin/v1/worker/attention_model_runner.py:290]: Done in 93cc04e. No longer contains the DBO workaround or its generic tests/docs; the compatibility fix is isolated in #217.

**Comment by sinpeyw** [comment]
### Qwen3-30B-A3B CUDA AFD throughput comparison

Following up on [the performance and A/F profiling question](https://github.com/vllm-project/afd-plugin/pull/216#issuecomment-5226319066), this update provides the end-to-end throughput comparison.

I reproduced the synchronous decode experiment from the [AFD performance blog](https://vllm.ai/blog/2026-07-23-vllm-afd-plugin) at an eight-GPU scale using Qwen3-30B-A3B on NVIDIA A800 GPUs with vLLM 0.26.0. The tested code combines #216, #217, and #219.

Native DP8/EP8 and 6A2F with A16/rank are the scaled counterparts of the published EP64 and 48A16F cases. They preserve the 3:1 Attention-to-FFN ratio, one-quarter active expert pool, per-Attention-rank batch sizes (16/28 for 16K and 8/14 for 32K), fixed input lengths, 512–1,536-token uniform outputs, forced balanced routing, and DBO for AFD. A32/rank and A64/rank are additional expert-pool sensitivity cases. The published 64A16F topology was not reproduced because its scaled 8A2F form requires ten GPUs.

Each case used 1,000 requests. Stable throughput is measured over the longest contiguous interval where running requests remain at least 90% of the run's peak, normalized by all eight deployed GPUs. All runs completed with zero failed requests.

| Topology | 16K tok/s/GPU | Change | 32K tok/s/GPU | Change |
| --- | ---: | ---: | ---: | ---: |
| Native DP8/EP8 | 504.21 | baseline | 255.91 | baseline |
| 6A2F, A16/rank | 524.88 | +4.10% | 269.08 | +5.14% |
| 6A2F, A32/rank | 525.30 | +4.18% | 269.18 | +5.19% |
| 6A2F, A64/rank | 373.24 | -25.97% | 189.92 | -25.79% |

image
image

Within this controlled setup, 6A2F with A16/rank and A32/rank consistently outperformed native DP8/EP8 at both input lengths: by 4.10–4.18% at 16K and 5.14–5.19% at 32K.

However, the A16/rank and A32/rank improvements are not sufficient evidence of a general AFD advantage because these AFD cases restrict the active expert pool. The A64/rank regression further illustrates this limitation. Broader topology studies with natural routing are still needed.

Once #216 is merged, I will follow up with a separate PR containing the detailed experimental setup and reproduction instructions.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        str(afd_port + case_index),
+        "--tp-size",
+        str(case.tp_size),
+        "--startup-timeout",
+        os.environ.get("AFD_GPU_E2E_STARTUP_TIMEOUT", "900"),
+        "--max-tokens",
+        str(case.max_tokens),
+        "--prompt",
+        DEFAULT_PROMPT,
+        "--num-requests",
```
Comment (yujuancao07) [inline tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: 2a2f_eager_smoke uses the default requests=1 and concurrency=1. With DP=2, a single request may only exercise one Attention–FFN pair, leaving the second pair untested. Please use at least two concurrent requests for the 2A2F cases.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    model = os.environ.get("AFD_QWEN3_MOE_E2E_MODEL")
+    if not model:
+        pytest.skip("set AFD_QWEN3_MOE_E2E_MODEL to a Qwen3 MoE model path")
+    return model
+
+
+def _gpu_list() -> list[str]:
+    return [
+        item.strip()
+        for item in os.environ.get("AFD_GPU_E2E_GPUS", "0,1,2,3").split(",")
```
Comment (yujuancao07) [inline tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: _gpu_list() defaults to physical GPUs 0,1,2,3 and ignores CUDA_VISIBLE_DEVICES.  Please require an explicit device list or derive it from CUDA_VISIBLE_DEVICES.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        requests=4,
+        concurrency=4,
+    ),
+)
+
+
+def _model_path() -> str:
+    model = os.environ.get("AFD_QWEN3_MOE_E2E_MODEL")
+    if not model:
+        pytest.skip("set AFD_QWEN3_MOE_E2E_MODEL to a Qwen3 MoE model path")
```
Comment (yujuancao07) [inline tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: If AFD_QWEN3_MOE_E2E_MODEL is missing, all Qwen3 tests are skipped and the job can still be green without testing Qwen3 at all.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            [
+                "--enable-dbo",
+                "--dbo-decode-token-threshold",
+                "1",
+                "--dbo-prefill-token-threshold",
+                "8",
+            ],
+        )
+
+    subprocess.run(command, cwd=REPO_ROOT, check=True)
```
Comment (yujuancao07) [inline tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: These cases only check whether the subprocess exits successfully. A silent fallback to eager execution, or DBO running without forming multiple microbatches, would still pass. Please add an assertion based on an explicit runtime signal that confirms Graph replay and DBO execution.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py (hunk 2 above)
Comment (sinpeyw) [reply tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: Addressed in a4de153. Every 2A2F case now uses at least four concurrent requests; the pressure case uses 64 requests at concurrency 32.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py (hunk 3 above)
Comment (sinpeyw) [reply tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: Addressed in a4de153. AFD_GPU_E2E_GPUS takes precedence, then CUDA_VISIBLE_DEVICES is used; the test fails if neither provides a device allocation.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py (hunk 4 above)
Comment (sinpeyw) [reply tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: Addressed in a4de153. Missing AFD_QWEN3_MOE_E2E_MODEL now fails the Qwen3 GPU suite instead of skipping it.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py (hunk 5 above)
Comment (sinpeyw) [reply tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: Addressed in a4de153. Graph cases assert a FULL runtime dispatch; the DBO case asserts runtime microbatching and exactly two UBatchSlice entries after API readiness.

File: afd_plugin/model_executor/models/qwen3_moe.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        # ### PATCH END: allocate only the active role's execution stage.
+
+    def compute_ffn_output(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        """Execute this layer's complete native FFN block on the FFN role."""
+        if self.afd_role != "ffn":
+            raise RuntimeError("Qwen3 MoE FFN compute requires the AFD FFN role")
+        return self.mlp(hidden_states)
+
+
+@native.support_torch_compile
```
Comment (hsliuustc0106) [inline afd_plugin/model_executor/models/qwen3_moe.py]: [P1] Avoid double compile initialization

Native `Qwen3MoeModel` is already wrapped by `support_torch_compile`, and this subclass calls that wrapped `super().__init__()`. Decorating the subclass again initializes `TorchCompileWithNoGuardsWrapper` twice in non-eager mode, creating duplicate compiler backends and bytecode hooks while overwriting the first cleanup handle. Please inherit the native compile wrapper without applying the decorator again, and add a regression test that compile support initializes once.

File: afd_plugin/model_executor/models/qwen3_moe.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+            yield name, loaded_weight
+
+
+def _validate_supported_config(
+    vllm_config: VllmConfig,
+    afd_config: AFDConfig,
+) -> None:
+    """Reject Qwen3 MoE modes not covered by the initial CUDA contract."""
+    parallel_config = vllm_config.parallel_config
+    if current_platform.device_type != "cuda":
```
Comment (hsliuustc0106) [inline afd_plugin/model_executor/models/qwen3_moe.py]: [P2] Distinguish CUDA from ROCm

vLLM's `RocmPlatform` also reports `device_type = "cuda"`, so this condition accepts the explicitly unsupported ROCm backend. The exact vLLM 0.26 validator accepted a `RocmPlatform` instance in a direct probe. Please use `if not current_platform.is_cuda():` and add a ROCm negative test.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        pytest.fail("GPU E2E device list must contain at least one device")
+    return gpus
+
+
+def _assert_runtime_signals(case: Qwen3MoeE2ECase, output: str) -> None:
+    _, ready, runtime_output = output.partition(API_READY_SIGNAL)
+    assert ready, "runner completed without reporting Attention API readiness"
+
+    if case.graph_capture_size:
+        assert GRAPH_RUNTIME_SIGNAL in runtime_output, (
```
Comment (hsliuustc0106) [inline tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: [P2] Verify FFN graph replay

This signal is emitted by the native Attention runner and does not prove that `GPUFFNModelRunner` selected `AFDGraphRunMode.REPLAY`. If the FFN graph key is absent, the FFN runner falls back to eager execution while this assertion still passes. Please emit an FFN-side replay signal and assert it after API readiness.

**Review by hsliuustc0106 — COMMENTED** [review]
Three inline findings on the current head: one compiler-lifecycle issue and two validation gaps.

**Comment by jiangkuaixue123** [comment]
Thanks for this PR.

Since the scope explicitly notes that this does not make any performance claims, I suggest we discuss the performance-related aspects of this change at our regular meeting. Please see the agenda for reference and add a topic/agenda item in the doc if you would like it discussed:

https://docs.google.com/document/d/1FMeSNdIyQEgeZUqj4ddalP__rZdzJSgzob1csEhPP3I

File: afd_plugin/model_executor/models/qwen3_moe.py (hunk 6 above)
Comment (sinpeyw) [reply afd_plugin/model_executor/models/qwen3_moe.py]: Fixed. Removed the duplicate decorator. The regression verifies that compile support comes from the native wrapper and is not applied again by the AFD subclass.

File: afd_plugin/model_executor/models/qwen3_moe.py (hunk 7 above)
Comment (sinpeyw) [reply afd_plugin/model_executor/models/qwen3_moe.py]: Fixed. The validator now uses current_platform.is_cuda(), with a negative test covering ROCm semantics.

File: tests/e2e/models/qwen3_moe/test_e2e_gpu.py (hunk 8 above)
Comment (sinpeyw) [reply tests/e2e/models/qwen3_moe/test_e2e_gpu.py]: Fixed. The FFN runner now emits a DEBUG signal only from AFDGraphRunMode.REPLAY, and the graph E2E asserts it after API readiness.

**Comment by sinpeyw** [comment]
Rebased onto the latest main. E2E changes were removed and will follow in a separate PR.

**Comment by jiangkuaixue123** [comment]
LGTM. Thanks for the contribution!
